File: src/Controllers/MatchesController.hs
{-# LANGUAGE DeriveGeneric #-}
{-# OPTIONS_GHC -Wno-missing-fields #-}
{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
{-# HLINT ignore "Redundant bracket" #-}

module Controllers.MatchesController where

import GHC.Generics
import Data.Aeson
import Controllers.BoardController;
import Controllers.AccountsController
import Controllers.PlayerController
import Controllers.LettersController
import Utils.Utils as UT
import Data.Time.Clock
import System.Console.ANSI



-- Uma partida é uma entidade que armazena todos os dados de um jogo, ele deve ter associado a ele:
-- Um nome de conta
-- Um tabuleiro onde será trabalhada a lógica de disposição, validação e pontuação de palavras
-- Um booleano que representará de quem é a vez de jogar, caso false, player 1, caso contrário, player 2
-- O jogador 1 que jogará na partida
-- O jogador 2 que jogará na partida
-- As letras que poderão ser distribuídas para os jogadores e que se acabarem acaba o jogo
-- As palavras que já foram dispostas no tabuleiro
-- O tempo que o jogador tem para terminar a jogada antes de passar para o próximo jogador
-- A quantidade de jogadas puladas no jogo, com limite de 4 até o jogo acabar
data Match = Match {
    mName :: String,
    mBoard :: Board,
    mTurn :: Bool,
    mP1 :: Player,
    mP2 :: Player,
    mLetters :: [Letter],
    mUsedWords :: [String],
    mTimer :: Float,
    mSkips :: Int
} deriving (Show, Generic, Eq)

instance ToJSON Match
instance FromJSON Match


-- Constante do path para as matches.json
-- Retorna: path do matches.json
matchesPath :: String
matchesPath = "data/matches.json"


-- Salva uma partida no json
-- Recebe: partida que será salva no json
saveMatchJson :: Match -> IO()
saveMatchJson match = UT.incJsonFile match matchesPath


-- Deleta uma partida do json
-- Recebe: partida que será deletada no json
deleteMatchFromJson :: String -> IO()
deleteMatchFromJson name = do
    match <- getMatchByName name
    UT.deleteFromJson match matchesPath


-- Atualiza uma partida do json
-- Recebe: partida que será atualizada no json
updateMatchJson :: Match -> IO()
updateMatchJson updatedMatch = do
    accs <- getMatches
    let updatedAccs = _getUpdatedMatches accs targetMatchName updatedMatch
    UT.writeJsonFile updatedAccs matchesPath
    where
        targetMatchName = mName updatedMatch


-- Cria uma partida e a retorna, adicionando-a no json
-- Recebe: nome da partida que será criada
-- Recebe: conta do player 1 que jogará a partida
-- Recebe: conta do player 2 que jogará a partida
-- Retorna: partida criada com as configuracoes iniciais
createMatch :: String -> Account -> Account -> IO Match
createMatch name acc1 acc2 = do
    let match = Match {
            mName = name,
            mBoard = startBoard,
            mTurn = False,
            mP1 = createPlayer acc1,
            mP2 = createPlayer acc2,
            mLetters = startLetters,
            mUsedWords = [],
            mTimer = 300,
            mSkips = 0
        }

    matchP1Letters <- updatePlayerLetters match
    let matchP2 = toggleMatchTurn matchP1Letters
    matchP2Letters <- updatePlayerLetters matchP2
    let initMatch = toggleMatchTurn matchP2Letters

    saveMatchJson initMatch
    return initMatch


-- Finaliza uma partida, apagando-a do json, atualizando a pontuacao das contas que jogaram
-- e retornando os status final da partida
-- Recebe: partida que será finalizada
-- Retorna: partida finalizada com seus status finais
finishMatch :: Match -> IO Match
finishMatch match = do
    incAccScore (accName (pAcc p1)) (pScore p1)
    incAccScore (accName (pAcc p2)) (pScore p2)
    deleteMatchFromJson (mName match)
    return match
    where
        p1 = mP1 match
        p2 = mP2 match


-- Atualiza o timer da partida
-- Recebe: partida que terá o timer atualizado
-- Recebe: tempo que o timer da partida terá
-- Retorna: partida com timer atualizado
updateMatchTimer :: Match -> Float -> Match
updateMatchTimer match time = match { mTimer = time }


-- Verifica se uma partida existe no json
-- Recebe: nome da partida que será verificada se existe
-- Retorna: bool que indicará se a partida existe ou nao 
matchExists :: String -> IO Bool
matchExists name = do
    maybeMatch <- getMatchByName name
    return $ case maybeMatch of
        Nothing -> False
        Just _ -> True


-- Retorna todas as partidas salvas no json
-- Retorna: array das partidas salvas no json
getMatches :: IO [Match]
getMatches = do
    UT.readJsonFile matchesPath


-- Retorna uma partida salva no json, procurando-a pelo nome
-- Recebe: nome da partida que será retornada
-- Retorna: partida que está armazenada no json, ou nada se a partida nao existir
getMatchByName :: String -> IO (Maybe Match)
getMatchByName targetName = do
    matches <- getMatches
    return $ UT.getObjByField matches mName targetName



-- Incrementa o score do player que está no turno
-- Recebe: partida que terá o player que terá score incrementado
-- Recebe: score que será incrementado no score do player
-- Retorna: partida com o score do player atualizado
incPlayerScore :: Match -> Int -> Match
incPlayerScore match score
    | mTurn match = match {mP2 = incScore (mP2 match) score}
    | otherwise = match {mP1 = incScore (mP1 match) score}


-- Atualiza as letras do player até ele ter 7 letras no máximo
-- Recebe: partida que terá as letras que serão dadas ao player
-- Retorna: partida com as letras do player e dela própria atualizadas
updatePlayerLetters :: Match -> IO Match
updatePlayerLetters match = do
    (playerLetters, updatedLetters) <- UT.popRandomElements (mLetters match) (7 - (length (pLetters player)))

    let updatedMatch = _updateMatchLetters match updatedLetters
    let updatedPlayer = addLetters player playerLetters

    return $ _updateMatchPlayer updatedMatch updatedPlayer
        where
            player
                | mTurn match = mP2 match
                | otherwise = mP1 match


-- Retorna o player com maior pontuação na partida
-- Recebe: partida que terá os players que terão o score comparados
-- Retorna: player com maior pontuação na partida
getBestPlayer :: Match -> Player
getBestPlayer match
    | p1Score > p2Score = p1
    | p1Score < p2Score = p2
    | p1Score == p2Score && p1RemainingScore < p2RemainingScore = p1
    | p1Score == p2Score && p2RemainingScore > p1RemainingScore = p2
    | otherwise = Player{pAcc = Account{accName = "Empate :D"}, pScore = p1Score}
    where 
        p1 = mP1 match
        p2 = mP2 match
        p1Score = pScore p1
        p2Score = pScore p2
        p1RemainingScore = getLetterSum $ pLetters p1
        p2RemainingScore = getLetterSum $ pLetters p2


-- Muda o turno da partida e soma 1 skip na contagem de skips da partida
-- Recebe: partida que terá o turno skipado
-- Retorna: partida com um skip a mais e rodada pro outro player
skipPlayerTurn :: Match -> Match
skipPlayerTurn match = (toggleMatchTurn match) {mSkips = mSkips match + 1}


-- Reseta a quantidade de skips da partida para 0
-- Recebe: partida que terá os skips resetados
-- Retorna: partida com os skips resetados para 0
resetMatchSkipsQtd :: Match -> Match
resetMatchSkipsQtd match = match{mSkips = 0}


-- Muda o turno da partida para o próximo player 
-- Recebe: partida que terá o turno mudado
-- Retorna: partida com o turno mudado
toggleMatchTurn :: Match -> Match
toggleMatchTurn match = match {mTurn = not (mTurn match), mTimer = 300}


-- Atualiza o tabuleiro da partida
-- Recebe: partida que terá o tabuleiro mudado
-- Recebe: tabuleiro atualizado que a partida possuirá
-- Retorna: partida com o tabuleiro atualizado
updateMatchBoard :: Match -> Board -> Match
updateMatchBoard match newBoard = match {mBoard = newBoard}


-- Atualiza as palavras usadas na partida
-- Recebe: partida que terá as palavras utilizadas atualizadas
-- Recebe: array de palavras que foram utilizadas na jogada
-- Retorna: partida com as palavras utilizadas atualizadas
updateMUsedWords :: Match -> [String] -> Match
updateMUsedWords match words = match {mUsedWords = words ++ mUsedWords match}


-- Troca uma letra selecionada pelo jogador por uma aleatória armazenada pela partida
-- Recebe: partida que terá as letras do player e dela mesma atualizadas
-- Recebe: letra que o jogador vai trocar com a partida
-- Retorna: partida com as letras do player e dela mesma atualizadas
switchPlayerLetter :: Match -> Letter -> IO Match
switchPlayerLetter match letter = do
    (newLetter, updatedLetters) <- UT.popRandomElements (mLetters match) 1

    let playerLetters = UT.removeOneElement (pLetters player) letter
    let updatedPlayer = updateLetters player playerLetters 

    let updatedMatch = _updateMatchLetters match (letter:updatedLetters)
    let updatedPlayer' = addLetters updatedPlayer newLetter

    return $ _updateMatchPlayer updatedMatch updatedPlayer'
    where
        player
            | mTurn match = mP2 match
            | otherwise = mP1 match


-- Remove as letras do jogador do turno, letras essas passadas na função
-- Recebe: partida que terá o player que terá as letras removidas
-- Recebe: array de letras que serão removidas do jogador
-- Retorna: partida com o player com as letras passadas removidas
removePlayerLetters :: Match -> [Char] -> Match
removePlayerLetters match toRemove = _updateMatchPlayer match (updateLetters player newLetters)
    where 
        newLetters = getLetterArray (UT.removeChars toRemove [letter l | l <- (pLetters player)])
        player
            | mTurn match = mP2 match
            | otherwise = mP1 match


-- Função interna que fará a atualização do player da vez
-- Recebe: partida que terá o player atualizado
-- Recebe: player atualizado que será passado para a partida
-- Retorna: partida com player atualizado
_updateMatchPlayer :: Match -> Player -> Match
_updateMatchPlayer match player
    | mTurn match = match {mP2 = player}
    | otherwise = match {mP1 = player}


-- Função interna que retorna um array das partidas atualizadas
-- Recebe: array de partidas 
-- Recebe: nome da partida que será atualizada
-- Recebe: partida atualizada que substituirá a desatualizada
-- Retorna: array de partidas atualizadas
_getUpdatedMatches :: [Match] -> String -> Match -> [Match]
_getUpdatedMatches [] _ _ = []
_getUpdatedMatches (match:tail) targetMatchName updatedMatch
    | mName match == targetMatchName = (updatedMatch:tail)
    | otherwise = (match:_getUpdatedMatches tail targetMatchName updatedMatch)


-- Função interna que atualiza as letras que a partida possui
-- Recebe: partida que terá as letras atualizadas
-- Recebe: letras que serão postas na partida atualizada
-- Retorna: partida com as letras atualizadas
_updateMatchLetters :: Match -> [Letter] -> Match
_updateMatchLetters match newLetters = match {mLetters = newLetters}

File: prolog/Controllers/matches_controller.pl

% Verifica se a partida selecionada existe
% Recebe: Nome da partida
match_exists(MatchName) :-
    current_predicate(match/9),
    match(MatchName, _, _, _, _, _, _, _, _), !.


% Retorna o átomo da partida no fomato match(NomeDaPartida, NomeDoBoardDaPartida, TurnoDaPartida, NomeDoPlayer1, NomeDoPlayer2, [LetrasDaPartida], [PalavrasUsadasNaPartida], TimerDaPartida, QuantidadeDeSkipsDaPartida)
% Recebe: Nome da partida
% Retorna: Átomo da partida
get_match(MatchName, Match) :- 
    match(MatchName, BoardName, MatchTurn, P1Name, P2Name, MatchLetters, MatchWords, MatchTimer, MatchSkips),
    Match = match(MatchName, BoardName, MatchTurn, P1Name, P2Name, MatchLetters, MatchWords, MatchTimer, MatchSkips).


% Retorna o nome do board da partida
% Recebe: Nome da partida
% Retorna: Nome do board da partida
get_match_board_name(MatchName, MatchBoardName) :-
    match(MatchName, MatchBoardName,_,_,_,_,_,_,_).


% Retorna o turno em que a partida está, false se for do player1, true para o player2
% Recebe: Nome da partida
% Retorna: Turno da partida
get_match_turn(MatchName, MatchTurn) :- 
    match(MatchName, _, MatchTurn, _, _, _, _, _, _).


% Retorna o nome do player1 da partida
% Recebe: Nome da partida
% Retorna: Nome do player1 da partida
get_match_p1_name(MatchName, P1Name) :- 
    match(MatchName, _, _, P1Name, _, _, _, _, _).


% Retorna o nome do player2 da partida
% Recebe: Nome da partida
% Retorna: Nome do player2 da partida
get_match_p2_name(MatchName, P2Name) :- 
    match(MatchName, _, _, _, P2Name, _, _, _, _).


% Retorna as letras da partida
% Recebe: Nome da partida
% Retorna: Letras da partida em formato de "array"
get_match_letters(MatchName, MatchLetters) :- 
    match(MatchName, _, _, _, _, MatchLetters, _, _, _).


% Retorna as palavras ja usadas na partida
% Recebe: Nome da partida
% Retorna: Palavras usadas na partida
get_match_words(MatchName, MatchWords) :- 
    match(MatchName, _, _, _, _, _, MatchWords, _, _).


% Retorna o timer da partida
% Recebe: Nome da partida
% Retorna: O timer da partida
get_match_timer(MatchName, MatchTimer) :- 
    match(MatchName, _, _, _, _, _, _, MatchTimer, _).


% Retorna a quantidade de skips da partida
% Recebe: Nome da partida
% Retorna: A quantidade de skips da partida
get_match_skips(MatchName, MatchSkips) :- 
    match(MatchName, _, _, _, _, _, _, _, MatchSkips).


% Retorna o nome do player da rodada
% Recebe: Nome da partida
% Retorna: Nome do player da rodada
get_turn_player_name(MatchName, PlayerName) :-
    get_match_turn(MatchName, MatchTurn),

    (MatchTurn -> 
        get_match_p2_name(MatchName, PlayerName);
        get_match_p1_name(MatchName, PlayerName)).


% Cria uma partida com as letras iniciais, cria os players, e o board, além de adicionar as letras iniciais pra cada jogador, adicionando à persistência
% Recebe: Nome da partida e dos players
create_match(MatchName, P1Name, P2Name) :- 
    \+ match_exists(MatchName),
    acc_exists(P1Name),
    acc_exists(P2Name),
    
    matches_path(MatchesPath),

    create_player(MatchName, P1Name),
    create_player(MatchName, P2Name),
    
    atom_concat(MatchName, board, BoardName),
    create_board(BoardName),

    start_letters(StartLetters),
    inc_fact_file(MatchesPath, match(MatchName, BoardName, false, P1Name, P2Name, StartLetters, [], 300, 0), match),
    update_player_letters(MatchName),
    toggle_player_turn(MatchName),
    update_player_letters(MatchName),
    toggle_player_turn(MatchName), !.


% Deleta uma partida selecionada pelo nome da persistencia e do jogo
% Recebe: Nome da partida a ser deletada
del_match(MatchName) :-
    matches_path(MatchesPath),
    players_path(PlayersPath),
    boards_path(BoardsPath),

    get_match(MatchName, Match),
    get_match_board_name(MatchName, BoardName),
    get_board(BoardName, Board),
    get_match_p1_name(MatchName, P1Name),
    get_match_p2_name(MatchName, P2Name),
    get_player(MatchName, P1Name, P1),
    get_player(MatchName, P2Name, P2),

    del_fact_file(PlayersPath, P1, player),
    del_fact_file(PlayersPath, P2, player),
    del_fact_file(MatchesPath, Match, match),
    del_fact_file(BoardsPath, Board, board).


% Finaliza uma partida, adicionando os pontos recebidos aos jogadores e apagando a partida da persistencia
% Recebe: Nome da partida a ser finalizada
finish_match(MatchName) :- 
    match(MatchName, _, _, P1Name, P2Name, _, _, _, _),
    
    get_player_score(MatchName, P1Name, P1Score),
    get_player_score(MatchName, P2Name, P2Score),
    inc_acc_score(P1Name, P1Score),
    inc_acc_score(P2Name, P2Score),

    del_match(MatchName),!.


% Atualiza o timer da partida
% Recebe: Nome da partida e o timer a ser armazenado por ela
update_match_timer(MatchName, NewTimer) :- 
    matches_path(MatchesPath),

    get_match(MatchName, Match),
    match(MatchName, BoardName, MatchTurn, P1Name, P2Name, MatchLetters, MatchWords, _, MatchSkips),

    update_fact_file(MatchesPath, Match, match(MatchName, BoardName, MatchTurn, P1Name, P2Name, MatchLetters, MatchWords, NewTimer, MatchSkips), match).


% Atualiza as letras da partida
% Recebe: Nome da partida e letras a serem armazenadas por ela
update_match_letters(MatchName, NewLetters) :-
    matches_path(MatchesPath),

    get_match(MatchName, Match),
    match(MatchName, BoardName, MatchTurn, P1Name, P2Name, _, MatchWords, MatchTimer, MatchSkips),

    update_fact_file(MatchesPath, Match, match(MatchName, BoardName, MatchTurn, P1Name, P2Name, NewLetters, MatchWords, MatchTimer, MatchSkips), match).


% Retorna todas as partidas registradas na persistência em formato de átomos
% Retorna: Todas as partidas registradas em formato de átomos
get_matches(Matches) :- 
    findall(match(MatchName, BoardName, MatchTurn, P1Name, P2Name, MatchLetters, MatchWords, MatchTimer, MatchSkips),
            match(MatchName, BoardName, MatchTurn, P1Name, P2Name, MatchLetters, MatchWords, MatchTimer, MatchSkips),
            Matches).


% Retorna os nomes de todas as partidas registradas na persistência
% Retorna: Todas as partidas registradas em formato de "array" de "strings"
get_match_names(MatchesNames) :-
    get_matches(Matches),
    get_match_names(Matches, MatchesNames).
get_match_names([], []).
get_match_names([H|T], MatchesNames) :-
    get_match_names(T, AnotherMatchesNames),
    H =.. [_|[MatchName |_]],
    MatchesNames = [MatchName|AnotherMatchesNames].


% Aumenta o score do player do turno da partida
% Recebe: Nome da partida e score a ser adicionado ao player
inc_player_score(MatchName, PlayerScore) :- 
    get_turn_player_name(MatchName, PlayerName),

    inc_score(MatchName, PlayerName, PlayerScore), !.


% Atualiza as letras do player da rodada, removendo-as das letras da partida e preenchendo as do player até darem 7 letras
% Recebe: Nome da partida para as letras serem atualizadas
update_player_letters(MatchName) :- 
    get_match_letters(MatchName, MatchLetters),
    get_turn_player_name(MatchName, PlayerName),

    get_player_letters(MatchName, PlayerName, PlayerLetters),
    length(PlayerLetters, PlayerLettersLength),
    LettersQuantity is 7 - PlayerLettersLength,

    pop_random_elements(MatchLetters, LettersQuantity, RemovedLetters, UpdatedLetters),

    update_match_letters(MatchName, UpdatedLetters),
    add_letters(MatchName, PlayerName, RemovedLetters).


% Pula o turno do player da vez, adicionando +1 a quantidade de skips da partida e passando para o turno do proximo jogador
% Recebe: Nome da partida a ter o turno pulado
skip_player_turn(MatchName) :- 
    matches_path(MatchesPath),
    
    get_match(MatchName, Match),
    match(MatchName, BoardName, MatchTurn, P1Name, P2Name, MatchLetters, MatchWords, MatchTimer, MatchSkips),
    
    NewMatchSkips is MatchSkips + 1,
    
    update_fact_file(MatchesPath, Match, match(MatchName, BoardName, MatchTurn, P1Name, P2Name, MatchLetters, MatchWords, MatchTimer, NewMatchSkips), match),
    toggle_player_turn(MatchName).


% Retorna a quantidade de skips da partida para 0
% Recebe: Nome da partida a ter os skips retornado para 0
reset_match_skips(MatchName) :- 
    matches_path(MatchesPath),
    
    get_match(MatchName, Match),
    match(MatchName, BoardName, MatchTurn, P1Name, P2Name, MatchLetters, MatchWords, MatchTimer, _),

    update_fact_file(MatchesPath, Match, match(MatchName, BoardName, MatchTurn, P1Name, P2Name, MatchLetters, MatchWords, MatchTimer, 0), match).


% Passa o turno da partida para o próximo jogador
% Recebe: Nome da partida a ter o turno passado
toggle_player_turn(MatchName) :- 
    matches_path(MatchesPath),

    get_match(MatchName, Match),
    match(MatchName, BoardName, MatchTurn, P1Name, P2Name, MatchLetters, MatchWords, _, MatchSkips),

    (MatchTurn -> 
        NewMatchTurn = false;
        NewMatchTurn = true),

    update_fact_file(MatchesPath, Match, match(MatchName, BoardName, NewMatchTurn, P1Name, P2Name, MatchLetters, MatchWords, 300, MatchSkips), match).


% Adiciona palavras as palavras usadas na partida
% Recebe: Nome da partida a ter as palavras adicionadas e as palavras a serem adicionadas
inc_match_used_words(MatchName, UsedWords) :- 
    matches_path(MatchesPath),

    get_match(MatchName, Match),
    match(MatchName, BoardName, MatchTurn, P1Name, P2Name, MatchLetters, MatchWords, MatchTimer, MatchSkips),
    
    append(MatchWords, UsedWords, NewMatchWords),

    update_fact_file(MatchesPath, Match, match(MatchName, BoardName, MatchTurn, P1Name, P2Name, MatchLetters, NewMatchWords, MatchTimer, MatchSkips), match).


% Troca a letra do player da rodada por alguma aleatoria da partida
% Recebe: Nome da partida a ter a letra trocada pelo player e a letra que foi trocada pelo player
switch_player_letter(MatchName, Letter) :- 
    get_match_letters(MatchName, MatchLetters),
    pop_random_elements(MatchLetters, 1, NewLetter, UpdatedLetters),

    get_turn_player_name(MatchName, PlayerName),

    get_player_letters(MatchName, PlayerName, PlayerLetters),
    selectchk(Letter, PlayerLetters, UpdatedPlayerLetters),

    append(UpdatedPlayerLetters, NewLetter, FinalPlayerLetters),
    append(UpdatedLetters, [Letter], FinalMatchLetters),
    
    update_match_letters(MatchName, FinalMatchLetters),
    update_letters(MatchName, PlayerName, FinalPlayerLetters).


% Remove as letras passadas do jogador do turno atual da partida
% Recebe: Nome da partida e as letras que serão removidas do player
remove_player_letters(MatchName, ToRemove) :-
    get_turn_player_name(MatchName, PlayerName),

    get_player_letters(MatchName, PlayerName, PlayerLetters),

    remove_elements(PlayerLetters, ToRemove, Updated),

    update_letters(MatchName, PlayerName, Updated), !.
